google/cloud/storage/examples/storage_object_samples.cc

```diff
@@ -535,13 +535,158 @@ void WriteObjectWithKmsKey(google::cloud::storage::Client client, int& argc,
+  auto destination_bucket_name = ConsumeArg(argc, argv);
+  auto destination_object_name = ConsumeArg(argc, argv);
+  //! [rewrite object]
+  namespace gcs = google::cloud::storage;
+  [](gcs::Client client, std::string source_bucket_name,
+     std::string source_object_name, std::string destination_bucket_name,
+     std::string destination_object_name) {
+    gcs::ObjectMetadata meta = client.RewriteObjectBlocking(
+        source_bucket_name, source_object_name, destination_bucket_name,
+        destination_object_name, gcs::ObjectMetadata());
+    std::cout << "Rewrote object " << destination_object_name
+              << " Metadata: " << meta << std::endl;
+  }
+  //! [rewrite object]
+  (std::move(client), source_bucket_name, source_object_name,
+   destination_bucket_name, destination_object_name);
+}
+
+void RewriteObjectNonBlocking(google::cloud::storage::Client client, int& argc,
+                              char* argv[]) {
+  if (argc != 5) {
+    throw Usage{
+        "rewrite-object-non-blocking <source-bucket-name> <source-object-name>"
+        " <destination-bucket-name> <destination-object-name>"};
+  }
+  auto source_bucket_name = ConsumeArg(argc, argv);
+  auto source_object_name = ConsumeArg(argc, argv);
+  auto destination_bucket_name = ConsumeArg(argc, argv);
+  auto destination_object_name = ConsumeArg(argc, argv);
+  //! [rewrite object non blocking]
+  namespace gcs = google::cloud::storage;
+  [](gcs::Client client, std::string source_bucket_name,
+     std::string source_object_name, std::string destination_bucket_name,
+     std::string destination_object_name) {
+    gcs::ObjectRewriter rewriter = client.RewriteObject(
+        source_bucket_name, source_object_name, destination_bucket_name,
+        destination_object_name, gcs::ObjectMetadata());
+    gcs::ObjectMetadata meta = rewriter.ResultWithProgressCallback(
+        [](gcs::RewriteProgress const& progress) {
+          std::cout << "Rewrote " << progress.total_bytes_rewritten << "/"
+                    << progress.object_size << std::endl;
+        });
+    std::cout << "Rewrote object " << destination_object_name
+              << " Metadata: " << meta << std::endl;
+  }
+  //! [rewrite object non blocking]
+  (std::move(client), source_bucket_name, source_object_name,
+   destination_bucket_name, destination_object_name);
+}
+
+void RewriteObjectToken(google::cloud::storage::Client client, int& argc,
+                        char* argv[]) {
+  if (argc != 5) {
+    throw Usage{
+        "rewrite-object-token <source-bucket-name> <source-object-name>"
+        " <destination-bucket-name> <destination-object-name>"};
+  }
+  auto source_bucket_name = ConsumeArg(argc, argv);
+  auto source_object_name = ConsumeArg(argc, argv);
+  auto destination_bucket_name = ConsumeArg(argc, argv);
+  auto destination_object_name = ConsumeArg(argc, argv);
+
+  // Create a 16 MiB object to copy from, otherwise the rewrite finishes too
```

Comment:
Creating a source object is a little misleading here. I think this should be moved into the shell script to create a file of a specific file size. Follow-up PR is also okay here.

Reply:
Fixed.